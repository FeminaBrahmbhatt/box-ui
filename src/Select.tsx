import React, {
  FC,
  ReactElement,
  isValidElement,
  cloneElement,
} from "react";
import { SelectProps } from "./@interfaces";
import {
  Listbox,
  ListboxButton,
  ListboxOption,
  ListboxOptions,
} from "@headlessui/react";
import classNames from "classnames";
import {FiCheck, FiChevronDown} from 'react-icons/fi'

export const Select: FC<SelectProps> = ({
  options,
  selectedOption,
  setSelectedOption,
  placeholder,
  LeadingIcon,
  width,
}) => {

  const ArrowDown = FiChevronDown as React.FC<any>;
  const Check = FiCheck as React.FC<any>;

  const setProps = (
    icon: ReactElement,
    iconSize: number,
    additionalClass: string
  ) => {
    if (
      isValidElement<{ size?: number; className?: string }>(icon) &&
      "size" in icon.props
    ) {
      return cloneElement(icon, {
        size: iconSize,
        className: classNames(icon.props.className, additionalClass),
      });
    }
    return icon;
  };

  return (
    <div className={classNames("relative inline-block", width)}>
      <Listbox value={selectedOption && selectedOption.value} onChange={setSelectedOption}>
        {({ open }) => (
          <>
            <ListboxButton className={classNames("shadow-sm flex items-center text-md border border-gray-300 dark:border-gray-500 h-11 px-3.5 rounded-lg bg-white dark:bg-gray-800",
                "focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 whitespace-nowrap",
                {
                  "text-gray-900 dark:text-white": selectedOption,
                  "text-gray-500 dark:text-gray-300": !selectedOption,
                },
                width,)}>
              {LeadingIcon && (
                <div className="w-5 h-5 mr-2 overflow-hidden">
                  {setProps(LeadingIcon, 20, classNames("text-gray-400"))}
                </div>
              )}
              {selectedOption ? selectedOption.label : placeholder}
              <ArrowDown
                size={20}
                className={classNames(
                  "text-gray-500 dark:text-gray-300 transform duration-100 ease-out",
                  {
                    "-rotate-180": open,
                    "ml-auto": width,
                    "ml-3.5": !width,
                  },
                )}
              />
            </ListboxButton>
            <ListboxOptions className="absolute z-10 inline-flex flex-col w-full bg-white border border-gray-300 rounded-lg shadow-lg top-13 dark:border-gray-500 dark:bg-gray-800">
              {options?.map((option, index) => (
                <ListboxOption as={React.Fragment} key={option.value} value={option.value}>
                    {({focus, selected}) => (
                        <li className={classNames(classNames(
                        "flex items-center pl-3.5 pr-3 justify-between h-11 text-gray-900 dark:text-white text-md cursor-pointer hover:bg-primary-25 dark:hover:bg-gray-100 dark:hover:bg-opacity-10 whitespace-nowrap",
                        {
                          "bg-primary-25 dark:bg-gray-100 dark:bg-opacity-10":
                            focus,
                          "rounded-t-lg": index === 0,
                          "rounded-b-lg": index === options.length - 1,
                        },
                      ))}>
                        <div className="flex items-center">{option.label}</div>
                        {selected && <Check className="ml-5 text-primary-600 dark:text-white" />}
                    </li>
                    )}
                    
                </ListboxOption>
              ))}
            </ListboxOptions>
          </>
        )}
      </Listbox>
    </div>
  );
};
