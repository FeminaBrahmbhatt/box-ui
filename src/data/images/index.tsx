export type CountryISO = "IN" | "BE" | "DE" | "FR" | "IT" | "NL" | "RU" | "US";

type image =  CountryISO ;

// Development images
// const devImages: Record<image, string> = {
//   logo: require("./logo.png"),
//   demoAvatar: require("./demoAvatar.png"),

//   IN: require("./in.svg"),
//   BE: require("./be.svg"),
//   DE: require("./de.svg"),
//   FR: require("./fr.svg"),
//   IT: require("./it.svg"),
//   NL: require("./nl.svg"),
//   RU: require("./ru.svg"),
//   US: require("./us.svg"),

// };

// Production images (CDN)
const prodImages: Record<image, string> = {
  IN: "https://kapowaz.github.io/circle-flags/flags/in.svg",
  BE: "https://kapowaz.github.io/circle-flags/flags/be.svg",
  DE: "https://kapowaz.github.io/circle-flags/flags/de.svg",
  FR: "https://kapowaz.github.io/circle-flags/flags/fr.svg",
  IT: "https://kapowaz.github.io/circle-flags/flags/it.svg",
  NL: "https://kapowaz.github.io/circle-flags/flags/nl.svg",
  RU: "https://kapowaz.github.io/circle-flags/flags/ru.svg",
  US: "https://kapowaz.github.io/circle-flags/flags/us.svg",
};

export const images = prodImages;
// export const images = process.env.NODE_ENV === "production" ? prodImages : devImages;
