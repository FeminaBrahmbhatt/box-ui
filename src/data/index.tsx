export * from "./colors";
export * from "./countries";
export * from "./prices";
export * from "./navItems";
export * from "./options";
