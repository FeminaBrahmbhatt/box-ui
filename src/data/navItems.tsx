import React from "react";
import { INavItem } from "../@interfaces";
import {
  FiArrowLeft,
  FiBarChart2,
  FiCheckSquare,
  FiHome,
  FiLayers,
  FiSettings,
  FiFlag,
  FiMapPin,
} from "react-icons/fi";

const Home = FiHome as React.FC<any>;
const Layers = FiLayers as React.FC<any>;
const Settings = FiSettings as React.FC<any>;
const Flag = FiFlag as React.FC<any>;
const MapPin = FiMapPin as React.FC<any>;
const CheckSquare = FiCheckSquare as React.FC<any>;
const BarChart = FiBarChart2 as React.FC<any>;
const ArrowLeft = FiArrowLeft as React.FC<any>;

export const navItemsTop: INavItem[] = [
  {
    order: 0,
    path: "home",
    label: "Home",
    icon: <Home />,
  },
  {
    order: 1,
    path: "dashboard",
    label: "Dashboard",
    icon: <BarChart />,
    subItems: [
      {
        order: 0,
        path: "analytics",
        label: "Analytics",
      },
      {
        order: 1,
        path: "apps",
        label: "Apps",
      },
      {
        order: 2,
        path: "learn",
        label: "Learn",
      },
    ],
  },
  {
    order: 2,
    path: "projects",
    label: "Projects",
    icon: <Layers />,
  },
  {
    order: 3,
    path: "tasks",
    label: "Tasks",
    icon: <CheckSquare />,
  },
  {
    order: 4,
    path: "reporting",
    label: "Reporting",
    icon: <Flag />,
  },
  {
    order: 5,
    path: "rentals",
    label: "Rentals",
    icon: <MapPin />,
    subItems: [
      {
        order: 0,
        path: "all",
        label: "All rentals",
      },
      {
        order: 1,
        path: "popular",
        label: "Popular",
      },
      {
        order: 2,
        path: "recent",
        label: "Recently added",
      },
    ],
  },
];

export const navItemsBottom: INavItem[] = [
  {
    order: 0,
    path: "settings",
    label: "Settings",
    icon: <Settings />,
  },
  {
    order: 1,
    label: "Close Navbar",
    icon: <ArrowLeft />,
    toggleSidebar: true,
  },
];
