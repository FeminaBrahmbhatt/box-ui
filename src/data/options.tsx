import React from 'react'
import { IButtonItem } from "../@interfaces";
import { FiCircle, FiList, FiGrid } from "react-icons/fi";


const Circle = FiCircle as React.FC<any>;

const List = FiList as React.FC<any>;

const Grid = FiGrid as React.FC<any>;

export const options1: IButtonItem[] = [
  { value: "Leading", content: "Leading" },
  { value: "Middle", content: "Middle" },
  { value: "Trailing", content: "Trailing" },
];

export const options2: IButtonItem[] = [
  {
    value: "first",
    content:  (
      <>
        <span><Circle size={20} className='mr-2'/></span>
        First
      </>
    ),
  },
  {
    value: "second",
    content: (
      <>
        <span><Circle size={20} className='mr-2'/></span>
        Second
      </>
    ),
  },
  {
    value: "third",
    content: (
      <>
        <span><Circle size={20} className='mr-2'/></span>
        Third
      </>
    ),
  },
];

export const options3: IButtonItem[] = [
  {value:'list', content:<List size={20}/>},
  {value:'grid', content:<Grid size={20}/>},
]
