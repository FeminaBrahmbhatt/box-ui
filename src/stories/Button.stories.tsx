import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "../Button";
import { ButtonProps } from "../@interfaces";
import { FiArrowLeft, FiArrowRight, FiStar } from "react-icons/fi";
import { StoryLayout } from "../StoryLayout";
import { IconWrapper } from "../utils/IconWrapper";

export default{
  title: "Example/Button",
  component: Button,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
} as Meta;
interface Props extends ButtonProps {
  darkMode: boolean;
}

type Story = StoryObj<Props>;

export const Default: Story = {
  args: {
    darkMode: false,
    children: "Button",
    size: 'md',
    variant: 'primary',
    disabled: false,
    LeadingIcon: undefined,
    TrailingIcon: undefined,
    IconOnly: undefined,
    className: ''
  },
  render: (args) => {
    return (
      <StoryLayout {...args} className="flex inline-flex flex-row space-x-2">
        <Button {...args}>Primary</Button>
      </StoryLayout>
    );
  },
};

export const Variants: Story = {
  args: {
    darkMode: false,
    variant: "primary",
  },
  render: (args) => {
    return (
      <StoryLayout {...args} className="flex inline-flex flex-row space-x-2">
        <Button variant="primary">Primary</Button>
        <Button variant="secondary">Secondary</Button>
        <Button variant="secondaryGray">Secondary Gray</Button>
        <Button variant="tertiary">Tertiary</Button>
        <Button variant="tertiaryGray">Tertiary Gray</Button>
      </StoryLayout>
    );
  },
};

export const Sizes: Story = {
  args: {
    darkMode: false,
    variant: "primary",
    size: "md",
  },
  render: (args) => {
    return (
      <StoryLayout {...args} className="flex inline-flex flex-row space-x-2">
        <Button size="sm">sm</Button>
        <Button size="md">md</Button>
        <Button size="lg">lg</Button>
        <Button size="xl">xl</Button>
        <Button size="2xl">2xl</Button>
      </StoryLayout>
    );
  },
};

export const TrailingIconButton: Story = {
  args: {
    darkMode: false,
    children: "Button",
    TrailingIcon: <IconWrapper Icon={FiArrowRight} data-testid="trailingIcon"/>,
  },  
  render: (args) => (
    <StoryLayout {...args}>
      <Button {...args} />
    </StoryLayout>
  ),
};

export const LeadingIconButton: Story = {
  args: {
    darkMode: false,
    children: "Button",
    LeadingIcon: <IconWrapper Icon={FiArrowLeft} data-testid="LeadingIcon"/>,
  },
  render: (args) => (
    <StoryLayout {...args}>
      <Button {...args} />
    </StoryLayout>
  ),
};

export const IconOnlyButton: Story = {
  args: {
    darkMode: false,
    IconOnly: <IconWrapper Icon={FiStar} data-testid="iconOnly"/>,
  },
  render: (args) => (
    <StoryLayout {...args}>
      <Button {...args} />
    </StoryLayout>
  ),
};
