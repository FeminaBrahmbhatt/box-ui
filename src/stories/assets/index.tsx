
export type CountryISO = "BE" | "DE" | "FR" | "IT" | "NL" | "RU" | "US";

type image = CountryISO;

// Development images
const devImages: Record<string, image> = {
  NL: require("./nl.svg")
};

// Production images (CDN)
// const prodImages: Record<string, image> = {
//   NL: "https://res.cloudinary.com/tailwindcss/image/upload/v1635279277/nl_tpy2ab.svg",
// };

export const assets = devImages;