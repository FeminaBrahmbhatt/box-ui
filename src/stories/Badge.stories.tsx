import React from 'react';
import {BadgeProps} from '../@interfaces';
import { Badge } from '../Badge';
import {images} from '../data/images'
import {Meta, StoryObj} from '@storybook/react';
import { StoryLayout } from '../StoryLayout';
import { FiArrowRight, FiStar } from 'react-icons/fi';
import { IconWrapper } from '../utils/IconWrapper';

export default {
    title: 'Example/Badge',
    component: Badge,
    parameters: {
        layout: 'centered',
    },
    tags: ['autodocs'],
} as Meta;
interface Props extends BadgeProps {
    darkMode: boolean;
}

export const Default:StoryObj<Props> = {
    args: {
        darkMode: false,
        variant: "success",
        size: "md",
        children:<span>Label</span>
    }
}

export const LeadingIconBadge:StoryObj<Props>  = {
    args: {
        darkMode: false,
        variant: "success",
        size: "md",
        LeadingIcon: <IconWrapper Icon={FiStar} data-testid="leadingIcon"/>
    },
    render: (args) => {
        return (
            <StoryLayout {...args} className="inline-flex flex-col space-y-2">
                <Badge {...args}>
                    Label
                </Badge>
            </StoryLayout>
        )
    }
}

export const ImageIconBadge:StoryObj<Props>  = {
    args: {
        darkMode: false,
        variant: "gray",
        size: "md",
        LeadingIcon: <img src={images.NL} alt="nl" className="w-4 h-4" data-testid="iconOnly" />
    },
    render: (args) => {
        return (
            <StoryLayout {...args} className="inline-flex flex-col space-y-2">
                <Badge {...args}>
                    Newzeland
                </Badge>
            </StoryLayout>
        )
    }
}

export const TrailingIconBadge:StoryObj<Props>  = {
    args: {
        darkMode: false,
        variant: "primary",
        size: "md",
        TrailingIcon: <IconWrapper Icon={FiArrowRight} data-testid="trailingIcon"/>
    },
    render: (args) => {
        return (
            <StoryLayout {...args} className="inline-flex flex-col space-y-2">
                <Badge {...args}>
                    Label
                </Badge>
            </StoryLayout>
        )
    }
}

