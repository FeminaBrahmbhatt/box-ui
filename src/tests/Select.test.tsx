import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { composeStories } from '@storybook/react';
import * as stories from '../stories/Select.stories';
import { prices, countries } from '../data';

const { Default, SelectWithIcon } = composeStories(stories);

describe('ðŸ§ª Select Story Tests (Direct function call style)', () => {
  
  test('Default story renders with a default country and placeholder', () => {
    const Story = Default;
    const { container } = render(Story(Story.args ?? {}));

    // Check if the default country is visible
    expect(screen.getByText('Ahmedabad, IN')).toBeInTheDocument();
    // Snapshot test
    expect(container).toMatchSnapshot();
  });

  // Test case 2: SelectWithIcon story renders with a leading dollar sign icon and price
  test('SelectWithIcon story renders with a leading dollar sign icon and price', () => {
    const Story = SelectWithIcon;
    const { container } = render(Story(Story.args ?? {}));

    // Check if the icon is visible (using a test ID or a class, depending on how the icon is rendered)
    expect(screen.getByTestId('leading-icon')).toBeInTheDocument(); // If you've added a data-testid to the icon
    // Check if the default price is visible
    expect(screen.getByText('Any price')).toBeInTheDocument();
    // Snapshot test
    expect(container).toMatchSnapshot();
  });

  // Test case 3: Typing works in Default Select (Country selection)
  test('typing works in Default Select (Country selection)', async () => {
    const Story = Default;
    render(Story(Story.args ?? {}));
  
    const selectButton = screen.getByRole('button', { name: /Ahmedabad, IN/i });
    expect(selectButton).toBeInTheDocument();
    fireEvent.click(selectButton);
  
    const option = await screen.findByText('Amsterdam, NL');
  
    expect(option).toHaveTextContent('Amsterdam, NL');
  });  

  test('typing works in SelectWithIcon (Price selection)', async() => {
    const Story = SelectWithIcon;
    render(Story(Story.args ?? {}));

    const priceSelect = screen.getByRole('button', { name: /any price/i });
    expect(priceSelect).toBeInTheDocument();
    fireEvent.click(priceSelect);

    const option = await screen.findByText('100 - 200');
    fireEvent.click(option);
    
    expect(priceSelect).toHaveTextContent('100 - 200');
  });
});
 