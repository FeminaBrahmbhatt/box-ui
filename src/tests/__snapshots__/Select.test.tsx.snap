// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ§ª Select Story Tests (Direct function call style) Default story renders with a default country and placeholder 1`] = `
<div>
  <div
    class="-m-4"
  >
    <div
      class="flex p-4"
    >
      <div
        class="relative inline-block w-50"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          class="shadow-sm flex items-center text-md border border-gray-300 dark:border-gray-500 h-11 px-3.5 rounded-lg bg-white dark:bg-gray-800 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 whitespace-nowrap text-gray-900 dark:text-white w-50"
          data-headlessui-state=""
          id="headlessui-listbox-button-:r2:"
          type="button"
        >
          <img
            alt="in"
            class="w-5 h-5 mr-2"
            src="https://kapowaz.github.io/circle-flags/flags/in.svg"
          />
           
          Ahmedabad, IN
          <svg
            class="text-gray-500 dark:text-gray-300 transform duration-100 ease-out ml-auto"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ðŸ§ª Select Story Tests (Direct function call style) SelectWithIcon story renders with a leading dollar sign icon and price 1`] = `
<div>
  <div
    class="-m-4"
  >
    <div
      class="flex p-4"
    >
      <div
        class="relative inline-block w-50"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          class="shadow-sm flex items-center text-md border border-gray-300 dark:border-gray-500 h-11 px-3.5 rounded-lg bg-white dark:bg-gray-800 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 whitespace-nowrap text-gray-900 dark:text-white w-50"
          data-headlessui-state=""
          id="headlessui-listbox-button-:r8:"
          type="button"
        >
          <div
            class="w-5 h-5 mr-2 overflow-hidden"
            data-testid="leading-icon"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="12"
                x2="12"
                y1="1"
                y2="23"
              />
              <path
                d="M17 5H9.5a3.5 3.5 0 0 0 0 7h5a3.5 3.5 0 0 1 0 7H6"
              />
            </svg>
          </div>
          Any price
          <svg
            class="text-gray-500 dark:text-gray-300 transform duration-100 ease-out ml-auto"
            fill="none"
            height="20"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
